src/combobox/ComboBox.ts

```diff
@@ -0,0 +1,389 @@
+import { WidgetBase, onPropertiesChanged } from '@dojo/widget-core/WidgetBase';
+import { ThemeableMixin, ThemeableProperties, theme } from '@dojo/widget-core/mixins/Themeable';
+import { v, w } from '@dojo/widget-core/d';
+import { DNode, PropertiesChangeEvent } from '@dojo/widget-core/interfaces';
+import FactoryRegistry from '@dojo/widget-core/FactoryRegistry';
+import { includes } from '@dojo/shim/array';
+import ResultMenu, { ResultMenuProperties } from './ResultMenu';
+import ResultItem from './ResultItem';
+import TextInput from '../textinput/TextInput';
+import Label, { LabelOptions } from '../label/Label';
+
+import * as css from './styles/comboBox.css';
+
+/**
+ * @type ComboBoxProperties
+ *
+ * Properties that can be set on a ComboBox component
+ *
+ * @property autoBlur			Determines whether the input should blur after value selection
+ * @property clearable			Determines whether the input should be able to be cleared
+ * @property customResultItem	Can be used to render a custom result
+ * @property customResultMenu	Can be used to render a custom result menu
+ * @property disabled			Prevents user interaction and styles content accordingly
+ * @property formId				ID of a form element associated with the form field
+ * @property inputProperties	HTML properties supported by FormLabelMixin to set on the underlying input
+ * @property invalid			Determines if this input is valid
+ * @property label				Label to show for this input
+ * @property openOnFocus		Determines whether the result list should open when the input is focused
+ * @property readOnly			Prevents user interaction
+ * @property required			Determines if this input is required, styles accordingly
+ * @property results			Results for the current search term; should be set in response to `onRequestResults`
+ * @property value				Value to set on the input
+ * @property getResultLabel		Can be used to get the text label of a result based on the underlying result object
+ * @property onBlur				Called when the input is blurred
+ * @property onChange			Called when the value changes
+ * @property onFocus			Called when the input is focused
+ * @property onRequestResults	Called when results are shown; should be used to set `results`
+ * @property onMenuChange		Called when menu visibility changes
+ */
+export interface ComboBoxProperties extends ThemeableProperties {
+	autoBlur?: boolean;
+	clearable?: boolean;
+	customResultItem?: any;
+	customResultMenu?: any;
+	disabled?: boolean;
+	formId?: string;
+	inputProperties?: any;
```

Comment:
I see that this is an object that contains a placeholder, the problem here is unless you have a custom diff it will not detect changes (as it is the same object reference) or if someone recreates the object but with the same placeholder value every time it will cause a property change to be detected.